### ISSUE TITLE
Feature: Exclusive mode (remove obsolete domains/zones)

### ISSUE THREAD

**Issue by Yannik** [StackExchange/dnscontrol#1976]
Currently, it is not possible to delete obsolete Domains/Zones by simply deleting them from the `dnsconfig.js` (I think).

It would be great if there was an `exclusive` Parameter that could be passed to `NewRegistrar` and `NewDnsProvider` that would allow removing all domains/zones not configured within dnscontrol.

**Comment by tlimoncelli** [comment]
Sadly we don't be adding this feature.

The risk of accidentally deleting a domain outweighs the benefit. 

That should be all the reason needed. However I'll also add:

* Most providers have different processes for deleting domains. Implementing this feature in a provider-agnostic way is difficult. The time spent designing this would outweigh the benefit.
* There's no harm in leaving a domain around unused.  If nothing delegates to it, the internet doesn't use it.
* DNSControl doesn't manage any registration or billing aspects of a domain.  If the domain needs to be deleted for billing reasons, this isn't the right tool.

Thank you for the suggestion but I have to reject this.

Tom

**Comment by Yannik** [comment]
Hi Tom,

IaC components should, in my opinion, provide the following benefits:

   -  idempotency,
   - diff support,
   - check mode compatibility,
   - correct change signaling, and
   - lifecycle management.

The last one, lifecycle management, is the one we are talking about here: Being able to delete outdated objects.

We have a customer with hundreds of domains for various projects, regional entities etc. Domains (and their respective zones!) are added and deleted all the time. Of course, managing deletes by hand is possible, but it is, depending on provider, quite a tedious process.

We are currently using a self-built ansible module to deal with all things domains & zones (including deletion ;-)), however we would very much like to switch to dnscontrol to benefit from the possibilities of a interpreted language instead of just static yaml. Unfortunately, it would really be a step back to losing lifecycle management features.

> The risk of accidentally deleting a domain outweighs the benefit.

For deleting domains I do agree that there is some risk involved. Which might outweigh the benefit, if you do not have a good change management process. 

For zones I cannot see a risk increase. Currently, you could accidentally delete all records by deleting them from the config, then you can delete the zone by accidentally deleting it from the config (and ignoring the preview!).  
To fix it: Just run `git reset --hard` and re-deploy.

To reduce risks further: Of course this would only be enabled with explicit configuration, and not by default. (Making it a default would be quite risky, I agree!)

I would like to add some more benefits to this discussion as well:  
Doing things by hand is prone to errors. Let's say a certain domain is not required anymore, and you want to delete it and all its services (including services hosted by third parties to which you CNAME). Regarding the lifecycle of the domain, you delete it in dnscontrol, but then forget to do the same in your domain/dns provider. The perfect recipe for subdomain hijacking.

> Most providers have different processes for deleting domains. Implementing this feature in a provider-agnostic way is difficult. The time spent designing this would outweigh the benefit.

The interfaces should, imo, be as simple as possible, e.g. `deleteDomain` and `deleteZone`. Implementation will be up to the providers and the requirements of their respective APIs. Additionally, we need a way to get all domains and zones. This already exists for zones and could, for most providers, easily be done for domains as well.

In conclusion: This is such a critical piece of the IaC-lifecycle, It would be great if you could reconsider this.

Best regards
Yannik

**Comment by Yannik** [comment]
Hi Tom,

not sure if you saw my reply because I didn't tag you. :)
Here you go: @tlimoncelli 

This is such a critical feature if you manage a lot of domains/zones as IaC, it would be great if you could reconsider it :)

**Comment by tlimoncelli** [comment]
Hi @Yannik 

This is a community-supported project. We all add the features that our organization needs.   I would welcome a PR that implements this feature.

I agree with most of your points, especially around the value of automation often being correctness especially for an error-prone thing like deleting domains.

However your assessment of the risk is not what I was thinking. Yes, you could re-add a zone by `Just run git reset --hard and re-deploy` if the zone was maintained by DNSControl in the past. My concern is about deleting a zone that wasn't maintained by DNSControl.  Imagine someone setting up DNSControl for their first time.  They own 5 domains.  As a test they set up DNSControl for one of those domains.  Their first "dnscontrol push" updates 1 zone and deletes the other 4.  `git --reset hard` isn't going to help them.

As far as the complexity of deleting a zone: Yes, "Implementation will be up to the providers" is ignoring some challenges. For example, name.com and gandi only let you delete a zone after it has expired. Should the failure to delete the zone be an error?  A warning?  If a zone isn't going to expire for a year, should the user suffer 12 months of error messages? I'm not saying this can't be solved, but it is something you'll need to work out.

Here's my suggestion:

* Start small.  For example, add a way to show which zones at a provider aren't mentioned in dnsconfig.js.  That will expose which domains should be deleted, or haven't been added to dnsconfig.js yet.  This could be its own subcommand (`dnscontrol list-unmanged-zones`) or a flag to `preview` and `push`.  (`--unmanaged-zones=list`).  That kind of feature will be appreciated by everyone in the community.  I know there have been times when I didn't realize  I had an unmanaged zone.
* Once that feature exists, add a way to act on that information.  A separate subcommand `dnscontrol delete-zone example.com` or  flag on preview/push (`--unmanaged-zone=delete`)

**Comment by Yannik** [comment]
Hi @tlimoncelli,

thanks for your feedback and for re-opening the issue!

I will see what we can do development-wise, and maybe other people are interested in this too. :)

**Comment by Yannik** [comment]
Hi @tlimoncelli 

I have added the `list-unmanaged-zones` command in #2026 and will continue to work on this in the following days.

**Comment by Yannik** [comment]
@tlimoncelli 

I have added a `list-unmanaged` command in #2028 and will continue to work on this :)

**Comment by tlimoncelli** [comment]
This is stale. Closing issue. Please re-open if needed.
